plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+
+from django.core.files import File
+from django.core.files.storage import default_storage
+from django.db.models import Q
+
+from pulpcore.plugin.models import Artifact, ProgressBar
+
+
+async def query_existing_artifacts(in_q, out_q):
+    """
+    Stages API stage that replaces :attr:`DeclarativeContent.content` objects with already-saved
+    :class:`~pulpcore.plugin.models.Artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    This stage inspects any unsaved :class:`~pulpcore.plugin.models.Artifact` objects and searches
+    using their metadata for existing saved :class:`~pulpcore.plugin.models.Artifact` objects inside
+    Pulp with the same digest value(s). Any existing :class:`~pulpcore.plugin.models.Artifact`
+    objects found will replace their unsaved counterpart in the
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The query_existing_artifacts stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content = []
```

Comment:
This name of `declarative_content` is a bit confusing.  I seems to be a *batch* of `DeclarativeContent`.  Having the *type* in the name implies that other variables like `content` are not `DeclarativeContent` and less obvious as to what the list is for.

Recommend renaming to: `batch` or something.

Reply:
batch sounds like a fine name. I will look into updating this name throughout the code.

Reply:
I updated with [this commit](https://github.com/pulp/pulp/pull/3559/commits/082e1e4888b65900e8fe6f1b20f252491ddc7deb). I used `batch` in almost all places except one where I thought `unhandled_content` was better since it's not guaranteed to handle the entire batch as a batch.